api/billingd/gateway/gateway.go

```diff
@@ -0,0 +1,193 @@
+			log.Fatalf("gateway error: %s", err)
+		}
+		log.Info("gateway was shutdown")
+	}()
+	log.Infof("gateway listening at %s", g.server.Addr)
+}
+
+// Addr returns the gateway's address.
+func (g *Gateway) Addr() string {
+	return g.server.Addr
+}
+
+// Stop the gateway.
+func (g *Gateway) Stop() error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	if err := g.server.Shutdown(ctx); err != nil {
+		return err
+	}
+	if err := g.client.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+const webhookMaxBodyBytes = int64(65536) // 64 KiB
+
+// webhookHandler handles stripe webhook events
+func (g *Gateway) webhookHandler(c *gin.Context) {
+	c.Request.Body = http.MaxBytesReader(c.Writer, c.Request.Body, webhookMaxBodyBytes)
+	payload, err := ioutil.ReadAll(c.Request.Body)
+	if err != nil {
+		log.Errorf("reading request body: %v", err)
+		c.Status(http.StatusBadRequest)
+		return
+	}
+
+	event := stripe.Event{}
+	if g.stripeWebhookSecret != "" {
+		event, err = webhook.ConstructEvent(payload, c.Request.Header.Get("Stripe-Signature"), g.stripeWebhookSecret)
+		if err != nil {
+			log.Errorf("verifying webhook signature: %v", err)
+			c.Status(http.StatusBadRequest)
+			return
+		}
+	}
+
+	switch event.Type {
+	case "customer.updated":
+		var cus stripe.Customer
+		if err := json.Unmarshal(event.Data.Raw, &cus); err != nil {
+			log.Errorf("parsing webhook JSON: %v", err)
+			c.Status(http.StatusBadRequest)
+			return
+		}
+		var billable bool
+		if !cus.Deleted {
+			if cus.InvoiceSettings.DefaultPaymentMethod != nil {
+				billable = true
+			} else if cus.DefaultSource != nil && !cus.DefaultSource.Deleted {
+				billable = true
+			}
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), handlerTimeout)
+		defer cancel()
+		if err := g.client.UpdateCustomer(ctx, cus.ID, cus.Balance, billable, cus.Delinquent); err != nil {
+			log.Errorf("updating customer: %v", err)
+			c.Status(http.StatusBadRequest)
+			return
+		}
+		log.Debugf("%s was updated", cus.ID)
+	case "customer.deleted":
+		// @todo: Does this need to be handled?
+	case "customer.subscription.updated",
+		"customer.subscription.deleted":
+		var sub stripe.Subscription
+		err := json.Unmarshal(event.Data.Raw, &sub)
+		if err != nil {
+			log.Errorf("parsing webhook JSON: %v", err)
+			c.Status(http.StatusBadRequest)
+			return
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), handlerTimeout)
+		defer cancel()
+		if err := g.client.UpdateCustomerSubscription(
+			ctx,
+			sub.Customer.ID,
+			sub.Status,
+			sub.CurrentPeriodStart,
+			sub.CurrentPeriodEnd,
+		); err != nil {
+			log.Errorf("updating customer subscription: %v", err)
+			c.Status(http.StatusBadRequest)
+			return
+		}
+		log.Debugf("%s subscription was updated", sub.ID)
+	default:
+		log.Debugf("unhandled event type: %s\n", event.Type)
+	}
+
+	c.Status(http.StatusOK)
```

Comment:
Yep, they do retry:
> In live mode, Stripe attempts to deliver your webhooks for up to three days with an exponential back off. In the Events section of the Dashboard, you can view when the next retry will occur.

More info [here](https://stripe.com/docs/webhooks/best-practices#events-and-retries). They also have this [event monitoring tool](https://github.com/stripe/stripe-webhook-monitor) that looks like it could be handy at some point.

Out event handling logic is idempotent, so duplicate events should not be a problem.